### PR TITLE
feat(pkg/trie): Add TrieCache for trie values and nodes

### LINKED ISSUE #3623: (pkg/trie): Implement trie nodes / values cache
In order to reduce the amount of memory used by our servers we have to implement a cache to prevent having 100% of the tries in memory.

This will be used with the lazy loading to have the most frequently used tries into the cached and improve the read times

### External resources:
[Substrate impl](https://github.com/paritytech/polkadot-sdk/blob/master/substrate/primitives/trie/src/cache/mod.rs#L18)

### PR DESCRIPTION
**Pull request by dimartiro**
## Changes

This PR introduces a `TrieCache` an LRU cache to reduce latency during trie lookups.

This implementation represents the simplest form of cache. In the future, we could consider a more complex one like parity does using different caches for different purposes, such as `SharedTrieCache`, `LocalTrieCache`, and `TrieCache`. 
These could serve as inspiration for future enhancements to our caching strategy. 
For further details, see [Parity's cache implementation](https://github.com/paritytech/polkadot-sdk/blob/master/substrate/primitives/trie/src/cache/mod.rs#L18-L35).

### Summary of changes

- [x] New `TrieCache` interface
- [x] In-memory trie cache implementation
    - [x] Tests
- [x] Start using `TrieCache` in `TrieDB`  
    - [x] Tests 
- [x] Add benchmarking without cache vs with cache

## Tests

```sh
make test
```

### Benchmarks

| Benchmark | Iterations | time / op | bytes / op | allocs / op |
|--------|--------:|--------:|--------:|--------:|
| get_value_without_cache-12 | 208468 | 5960 ns/op | 5609 B/op | 167 allocs/op |
| get_value_with_cache-12 | 8990666 | 132.4 ns/op | 116 B/op | 4 allocs/op |
| iterate_all_entries_without_cache-12 | 32265 | 37454 ns/op | 34783 B/op | 1053 allocs/op |
| iterate_all_entries_with_cache-12 | 127227 | 9784 ns/op | 7794 B/op | 280 allocs/op | 

## Issues

closes: #3623 

## Primary Reviewer

@EclesioMeloJunior

### CODE REVIEW

File: pkg/trie/triedb/codec/node.go [hunk 1]
```diff
@@ -48,6 +48,7 @@ type (
 	}
 	// HashedValue contains a hash used to lookup in db for real value
 	HashedValue struct {
+		// TODO: use common.Hash
```
Comment (jimjbrettj) [inline pkg/trie/triedb/codec/node.go]: still a todo?

File: pkg/trie/triedb/in_memory_to_triedb_migration_test.go [hunk 2]
```diff
@@ -8,20 +8,19 @@ import (
...
+	inmemory_trie "github.com/ChainSafe/gossamer/pkg/trie/inmemory"
 	"github.com/stretchr/testify/assert"
-	"github.com/stretchr/testify/require"
 )
 
-func newTestDB(t *testing.T) database.Table {
+func newTestDB(t assert.TestingT) database.Table {
 	db, err := database.NewPebble("", true)
-	require.NoError(t, err)
+	assert.NoError(t, err)
```
Comment (jimjbrettj) [inline pkg/trie/triedb/in_memory_to_triedb_migration_test.go:17]: why this change?

**Review by jimjbrettj — APPROVED** [review]
very nice change! good work

File: pkg/trie/triedb/triedb.go [hunk 3]
```diff
@@ -61,37 +64,30 @@ func (t *TrieDB) MustHash() common.Hash {
...
 	if err != nil {
 		return nil, err
 	}
 
-	return decodedNode, nil
+	reader := bytes.NewReader(encodedNode)
+	return codec.Decode(reader)
 }
 
+// Internal methods
```
Comment (EclesioMeloJunior) [inline pkg/trie/triedb/triedb.go:86]: not needed comment

File: pkg/trie/triedb/lookup.go [hunk 4]
```diff
@@ -103,14 +107,29 @@ func (l *TrieLookup) lookupNode(keyNibbles []byte) (codec.Node, error) {
...
 
-func (l *TrieLookup) lookupValue(keyNibbles []byte) ([]byte, error) {
+func (l *TrieLookup) lookupValue(keyNibbles []byte) (value []byte, err error) {
+	if l.cache != nil {
+		if value = l.cache.GetValue(keyNibbles); value != nil {
+			return value, nil
+		}
+	}
+
 	node, err := l.lookupNode(keyNibbles)
```
Comment (EclesioMeloJunior) [inline pkg/trie/triedb/lookup.go:129]: While looking up the node do you want to cache it as well?

File: pkg/trie/cache/inmemory/lru_cache.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package inmemory
+
+import (
+	"github.com/karlseguin/ccache/v3"
```
Comment (EclesioMeloJunior) [inline pkg/trie/cache/inmemory/lru_cache.go]: Gossamer currently have a LRU cache implementation at `lib/utils/lru-cache`, can it be used?

**Review by EclesioMeloJunior — CHANGES_REQUESTED** [review]

File: pkg/trie/cache/inmemory/lru_cache.go (hunk 5 above)
Comment (dimartiro) [reply pkg/trie/cache/inmemory/lru_cache.go]: Good catch, let me check, with that we can prevent adding new external deps! thks

File: pkg/trie/cache/inmemory/trie_cache.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+import (
+	"github.com/ChainSafe/gossamer/pkg/trie/cache"
+)
+
+// https://github.com/paritytech/polkadot-sdk/blob/a8f4f4f00f8fc0da512a09e1450bf4cda954d70d/substrate/primitives/trie/src/cache/mod.rs#L98
+const defaultNodeCacheMaxSize = 8 * 1024 * 1024  // 8MB
+const defaultValueCacheMaxSize = 2 * 1024 * 1024 // 2MB
+
+// TrieInMemoryCache is an in-memory cache for trie nodes
+type trieInMemoryCache struct {
```
Comment (timwu20) [inline pkg/trie/cache/inmemory/trie_cache.go]: Shouldn't this be public? Seems odd that the public constructor on line 21 returns a private type.

File: pkg/trie/cache/inmemory/lru_cache.go [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+)
+
+// ccache uses 350 bytes of overhead per entry
+const cacheValueOverheadSize = 350
```
Comment (timwu20) [inline pkg/trie/cache/inmemory/lru_cache.go]: where does this come from?

File: pkg/trie/cache/trie_cache.go [hunk 8]
```diff
@@ -0,0 +1,11 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package cache
+
+type TrieCache interface {
+	GetValue(key []byte) []byte
+	SetValue(key []byte, value []byte)
+	GetNode(key []byte) []byte
+	SetNode(key []byte, value []byte)
```
Comment (timwu20) [inline pkg/trie/cache/trie_cache.go]: Do we need to mirror the same interface as the parity trie-db `TrieCache` trait? I see that they don't have an analagous `SetNode` method, but rather have a  `get_or_insert_node` method.

File: pkg/trie/triedb/codec/node.go (hunk 1 above)
Comment (dimartiro) [reply pkg/trie/triedb/codec/node.go]: Yes but I'm addressing it in my next PR so I can remove it from here

File: pkg/trie/triedb/in_memory_to_triedb_migration_test.go (hunk 2 above)
Comment (dimartiro) [reply pkg/trie/triedb/in_memory_to_triedb_migration_test.go:17]: To remove the `require` dependency and use `assert` for every test

File: pkg/trie/cache/inmemory/lru_cache.go (hunk 5 above)
Comment (dimartiro) [reply pkg/trie/cache/inmemory/lru_cache.go]: The difference is the way it calculates the capacity, I need a "custom" capacity using the value size instead of the amount of entries. So I can't just replace it to use the one we already have 😞

File: pkg/trie/cache/inmemory/trie_cache.go (hunk 6 above)
Comment (dimartiro) [reply pkg/trie/cache/inmemory/trie_cache.go]: True! my bad. Fixed now!

File: pkg/trie/cache/inmemory/lru_cache.go (hunk 7 above)
Comment (dimartiro) [reply pkg/trie/cache/inmemory/lru_cache.go]: From `ccache` [doc](https://github.com/karlseguin/ccache?tab=readme-ov-file#size)

File: pkg/trie/cache/trie_cache.go (hunk 8 above)
Comment (dimartiro) [reply pkg/trie/cache/trie_cache.go]: We can do that, but I want to keep it simple. The `get_or_insert_node` function requires a callback function if the value isn't present. I prefer the Go way, which is more verbose, and explicitly call the `SetNode` method if we need to add that node.

File: pkg/trie/triedb/lookup.go (hunk 4 above)
Comment (dimartiro) [reply pkg/trie/triedb/lookup.go:129]: Done! thanks for raising that

File: pkg/trie/cache/inmemory/lru_cache.go (hunk 5 above)
Comment (EclesioMeloJunior) [reply pkg/trie/cache/inmemory/lru_cache.go]: I see, yeah, I don't mind to keep using ccache! Thanks for checking 👍

**Review by EclesioMeloJunior — APPROVED** [review]
LGTM

File: pkg/trie/cache/inmemory/trie_cache.go [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+package inmemory
+
+import (
+	"github.com/ChainSafe/gossamer/pkg/trie/cache"
+)
+
+// https://github.com/paritytech/polkadot-sdk/blob/a8f4f4f00f8fc0da512a09e1450bf4cda954d70d/substrate/primitives/trie/src/cache/mod.rs#L98
+const defaultNodeCacheMaxSize = 8 * 1024 * 1024  // 8MB
+const defaultValueCacheMaxSize = 2 * 1024 * 1024 // 2MB
```
Comment (timwu20) [inline pkg/trie/cache/inmemory/trie_cache.go]: I think it would be fine to use `ccache` to track number of items instead of bytes.  I don't think we need to adhere to the same strict size requirements for the node and value caches.  That would mean we wouldn't call `cacheValue.Size()` as often as well.

Comment (dimartiro) [reply pkg/trie/cache/inmemory/trie_cache.go]: If that's the case, we can reuse our current LRU cache implementation that @EclesioMeloJunior [mentioned](https://github.com/ChainSafe/gossamer/pull/3895#discussion_r1594361263) a few days ago. 
My only concern is that we don’t know what that best max elements number could be and the size could vary. By keeping our current implementation, we can ensure we are not using more than `maxSize` bytes and remain "compatible" with the parity implementation.

Comment (timwu20) [reply pkg/trie/cache/inmemory/trie_cache.go]: I think the `nodeCache` could definitely be counted by actual nodes instead of bytes and avoid doing the size checks.  For the `valueCache` it makes sense to track bytes and enforce a max limit.  I think this is a worthwhile change which differentiates our trie cache implementation for good reasons.

File: pkg/trie/cache/inmemory/lru_cache.go [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+// Size returns the size of the cacheValue taking the overhead into account
+func (cv cacheValue) Size() int64 {
+	return int64(len(cv) + cacheValueOverheadSize)
+}
+
+// lruCache is an in-memory lru cache
+// consider that the values are deleted asyncronously so there is a chance that
+// the maxSize can be exceeded
+// we can use lru.GC() to force the deletion of the items that should be deleted
+type lruCache struct {
```
Comment (timwu20) [inline pkg/trie/cache/inmemory/lru_cache.go]: differentiate what the difference is between this and the imported `LRUCache`?
```suggestion
type maxBytesLRUCache struct {
```

Comment (dimartiro) [reply pkg/trie/cache/inmemory/lru_cache.go]: Done!

**Review by timwu20 — APPROVED** [review]
